### PR TITLE
[SofaKernel] Activates thread-safetiness on MessageDispatcher.

### PR DESCRIPTION
**Pull request by damienmarchal**
Problem:
MessageDispatcher is not thread safe as showned with the failing test named 
testThreading_OpenIssue.

This PR:
    - the MessageDispatcher is now thread safe so we can emit messages from different threads. 
    - the test is not an OpenIssue anymore and its future failure should be considered as regression.

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed 
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
I have added #ifdef to activate/deactivate the mutex on demand. 

@matthieu-nesme I measured the performance with the mutex. On my machine, the message API is handling around 900000 messages per seconds. I got no noticeable difference when the mutex is there or not. Can someone confirm that ?

File: SofaKernel/framework/framework_test/helper/logging/logging_test.cpp [hunk 1]
```diff
@@ -162,8 +162,14 @@ void f3()
 }
 
 
-TEST(LoggingTest, threadingTests_OpenIssue)
+
+TEST(LoggingTest, threadingTests)
 {
+    if(!SOFA_WITH_THREADING){
+        std::cout << "Test canceled because sofa is not compiled with SOFA_WITH_THREADING option." << std::endl ;
```
Comment (matthieu-nesme) [inline SofaKernel/framework/framework_test/helper/logging/logging_test.cpp:169]: std::cout ??? :P

File: CMakeLists.txt [hunk 2]
```diff
@@ -225,6 +225,8 @@ option(SOFA_USE_MASK "Use mask optimization" ON)
 ### SOFA_DEV_TOOL
 option(SOFA_WITH_DEVTOOLS "Compile with developement extra features." ON)
 
+### Threading
+option(SOFA_WITH_THREADING "Compile sofa with thread-safetiness support (PARTIAL/EXPERIMENTAL)" ON)
```
Comment (matthieu-nesme) [inline CMakeLists.txt:229]: Safeness

File: SofaKernel/framework/framework_test/helper/logging/logging_test.cpp (hunk 1 above)
Comment (damienmarchal) [reply SofaKernel/framework/framework_test/helper/logging/logging_test.cpp:169]: :)
I hesitated but decided that in Logging_test I shouldn't make the assumption that msg_* is working.

**Comment by guparan** [comment]
@damienmarchal 
Could you resolve the conflicts (due to the merge of #213) in MessageDispatcher please?

**Comment by damienmarchal** [comment]
[ci-build]

**Comment by damienmarchal** [comment]
[ci-build]

**Pull request merged**